### ISSUE TITLE
cleanup tuple literals

### ISSUE THREAD

**Issue by mrjbq7** [slavapestov/factor#1449]
It's really ugly that tuple literals have a ``f`` to dispatch on whether they are boa or named:

  * ``T{ foo f 1 2 3 }`` becomes ``T{ foo 1 2 3 }`` or ``T[ foo 1 2 3 ]``
  * ``T{ foo { a 1 } { b 2 } { c 3 } }`` maybe becomes ``T{ foo a: 1 b: 2 c: 3 }``
  * (``T[`` syntax conflicts with ``foo[`` almost always being a typed quotation)

**Comment by mrjbq7** [comment]
This implements ``T[``:

```factor
SYNTAX: T[ scan-word \ ] parse-until boa>object suffix! ;
```

**Comment by mrjbq7** [comment]
```
T{ foo 1 2 3 }
T{ foo a: 1 b: 2 c: 3 }
```

**Comment by erg** [comment]
How about ``T{ foo 1 2 3 }`` for boa, then ``TA{ foo a: 1 b: 2 c: 3 }`` and ``TA{ foo { a 1 } { b 2 } { c 3 } }`` for associative array initialization?

**Comment by mrjbq7** [comment]
That's confusing maybe with `tuple-arrays`?

Also I kinda liked `T[` for some reason.

**Comment by erg** [comment]
``TH{`` then? The plan is to make all data behave like ``$`` and ``T[`` wouldn't really make sense anymore.

**Comment by mrjbq7** [comment]
That reminds me, I prefer `hashtable{` to `H{`.

**Comment by erg** [comment]
Yes, want to add that? We can also have ``array{`` ``vector{`` at the same time as ``{`` and ``V{``.

**Comment by mrjbq7** [comment]
Maybe `tuple{ foo 1 2 3 }` and `foo{ { a 1 } { b 2 } { c 3 } }` or `foo{ a: 1 b: 2 c: 3 }`.

**Comment by erg** [comment]
Oh, that's the good one. Yeah! (I have tons of notes about this stuff but forget!)

Can we clean that up for .98?

**Comment by erg** [comment]
I have a patch that adds ``TH{`` and ``foo{`` but prints them as ``T{ foo f 1 2 3 }``. Maybe it's still good enough to push?

```factor
diff --git a/core/classes/tuple/parser/parser.factor b/core/classes/tuple/parser/parser.factor
index 202214770b..e771595cf7 100644
--- a/core/classes/tuple/parser/parser.factor
+++ b/core/classes/tuple/parser/parser.factor
@@ -115,5 +115,18 @@ M: tuple-class boa>object
         [ bad-literal-tuple ]
     } case ;
 
+: parse-tuple-hash-literal-slots ( class slots -- tuple )
+    scan-token {
+        { "{" [ 2dup parse-slot-values assoc>object ] }
+        { "}" [ drop new ] }
+        [ bad-literal-tuple ]
+    } case ;
+
 : parse-tuple-literal ( -- tuple )
     scan-word dup all-slots parse-tuple-literal-slots ;
+
+: parse-tuple-hash-literal ( -- tuple )
+    scan-word dup all-slots parse-tuple-hash-literal-slots ;
+
+: parse-tuple-class-literal ( class -- tuple )
+    dup all-slots parse-tuple-hash-literal-slots ;
\ No newline at end of file
diff --git a/core/syntax/syntax.factor b/core/syntax/syntax.factor
index a0908b7040..0570b9b3ea 100644
--- a/core/syntax/syntax.factor
+++ b/core/syntax/syntax.factor
@@ -36,6 +36,10 @@ IN: bootstrap.syntax
     ] dip
     define-syntax ;
 
+: define-tuple-syntax-word ( class -- )
+    [ name>> "{" append create-word-in ] [ ] bi
+    [ parse-tuple-class-literal suffix! ] curry define-syntax ;
+
 [
     { "]" "}" ";" ">>" } [ define-delimiter ] each
 
@@ -105,6 +109,7 @@ IN: bootstrap.syntax
     "BV{" [ \ } [ >byte-vector ] parse-literal ] define-core-syntax
     "H{" [ \ } [ parse-hashtable ] parse-literal ] define-core-syntax
     "T{" [ parse-tuple-literal suffix! ] define-core-syntax
+    "TH{" [ parse-tuple-hash-literal suffix! ] define-core-syntax
     "W{" [ \ } [ first <wrapper> ] parse-literal ] define-core-syntax
     "HS{" [ \ } [ >hash-set ] parse-literal ] define-core-syntax
 
@@ -210,7 +215,7 @@ IN: bootstrap.syntax
     ] define-core-syntax
 
     "TUPLE:" [
-        parse-tuple-definition define-tuple-class
+        parse-tuple-definition [ define-tuple-class ] [ 2drop define-tuple-syntax-word ] 3bi
     ] define-core-syntax
 
     "final" [
```

**Comment by erg** [comment]
This patch doesn't bootstrap because ``maybe{`` is defined twice. I think we could add ``TH{`` right now and add ``foo{`` with the new parser as an op-in syntax so you can override it.

Something like this for the new parser:
```factor
TUPLE: foo a b ;
DEFAULT-DATA: \ foo{

TUPLE: maybe { class classoid read-only initial: object } ;
DATA: \ maybe{ <anonymous-union> <maybe> ;
```

**Comment by mrjbq7** [comment]
I'm kinda not a fan of TH.

**Comment by bpollack** [comment]
I'm just saying we are taking like _zero_ advantage of Factor's excellent Unicode support. `foo👉👈` collides with *nothing* in the default vocab.

More serious comment: I'm also not a huge fan of `TH{`. I would, however, be a fan of either just fixing the `maybe{` issue, or alternatively of shoving them into their own namespace for the moment to require opting in.

**Comment by erg** [comment]
Do we want to have some kind of PR/code review system so that I don't go pushing things yall don't like?

If we had it auto-test ``core`` vocabs with TravisCI then we could get good turnaround times, like 10m  and that would encourage people to let others comment on the patches as well. I'm not necessarily in favor of this, but it's possible to implement.

**Comment by bpollack** [comment]
Hmm, maybe. I know we had issues with Travis just taking too damn long, but maybe if we made it for a narrower part of core?

**Comment by bjourne** [comment]
Can't we just ditch the `T{ foo f 1 2 3 }` literal syntax? We already have `T{ foo { k v } { k2 v2 } ... }` for creating tuple literals.

**Comment by AlexIljin** [comment]
If anyone cares about my opinion, I would prefer to have `T{ foo a: 1 b: 2 c: 3 }` and ditch both `T{ foo f 1 2 3 }` (`f` is weird and the slot order is fragile) and `T{ foo { a 1 } { b 2 } { c 3 } }` (brace overload, difficult to read and type).

I bit more natural text-wise would actually be something like `T{ foo: a 1, b 2, c 3 }`.

**Comment by mrjbq7** [comment]
@AlexIljin we definitely care, we would like to keep it homoiconic as much as possible, and (perhaps) consistent with other "key/value" objects like hashtables.  I'm not a big fan of the comma-delimited syntax, but could see an implicit string syntax like your first example.  We've even talked about ditching ``T{ foo ... }`` in favor of ``foo{ ... }``...

**Comment by AlexIljin** [comment]
@mrjbq7: I'm glad to hear that, and I agree that comma-delimited lists aren't very Factor-y at the moment (in the current version of the syntax). In particular, they wouldn't work with the numbers: `T{ foo: a 1,000, ... }`. Maybe semicolons instead: `T{ foo: a 1,000,000; b 2,000,000 }`?

I like the `foo{ ... }` syntax, it seems appropriate, and feels like the same kind of thing that happens when you declare slots. You just get a few automatically generated service words, like `slot>>` and `foo{`. But would `foo{` work in the same file where `foo` is declared?

**Comment by mrjbq7** [comment]
Yep, it would need to work automatically.  Some features of the new lexer might enable this, perhaps.  It's important that whatever changes we make keeps the (relative) simplicity of the current way that syntax words work.

**Comment by nomennescio** [comment]
Warning, long comment ahead!

Factor's literal syntax is one of the better language features compared to other languages, with a pedigree in Smalltalk. The `T{` syntax is the odd one out, having three different parsing alternative combined into one "type prefix" (and similarly the `S{` literal). The `f` construct is not only ugly, but confusing, and blocks the 'flow' when reading code, and therefore should go. The current design of (data) literals mark everything inside a `{` `}` pair as content of the literal, which type is marked with zero or more distinguishing capitals followed by `{`, to make literals as compact as possible. No commas or other unneccessary syntactic fluff, terseness IS important, I therefore am not a fan of more verbose "type markers" before the `{`. New literal syntax rarely needs to be added, and as many concepts in Factor are modeled as classes/tuples, the `T{` syntax is actually very much the 'core' literal of the language, next to `[` for quotation literals ("code"), and `{` for array literals ("data"). By convention, the `{` marks a data literal, so in my opinion that should stay for a tuple literal. I think most would agree that the `T{` syntax for the empty tuple and for the named slots with an assoc is OK-ish, and apart from the `f`, the unnamed slot syntax is OK too.

Can we solve this then within these boundary conditions?

First of all, could we just drop the requirement of the `f`, at the expense of not generating a bad-literal-tuple error anymore, by changing the syntax parsing to:
```
: parse-tuple-literal-slots ( class slots -- tuple )
    scan-token {
        { "{" [ 2dup parse-slot-values assoc>object ] }
        { "}" [ drop new ] }
        [ \ } parse-until boa>object ]
    } case ;
```
In this case, unnamed slots requiring an array literal as first argument are out of luck, but that is not as bad as it sounds.

As an alternative, to have array literals as first argument to unnamed slots, some already mentioned changing the syntax for named slots, but the proposed syntax (`fieldname:`) maybe clashes with (future) typed variable/slot lexing support? Maybe another syntax could be used if necessary? For instance, `value @fieldname`, which could be even be mixed with unnamed slots. Having `@` as prefix requires special treatment of all names with that character as prefix, alternatively one could use `value @ fieldname` or `value fieldname @`.

Another alternative is to move one of the three parsing alternatives into a new literal syntax.

`N{` could be used for the option to have a 'N'ew literal for empty tuples. That would still give the issue with an array literal as first argument to `T{`. That leaves us with the last option of introducing a new literal syntax for unnamed field initialization. `T[` was already mentioned, but that suggests too much being a 'code' literal for my taste, alternatives could be `O{` ('object'), `F{` ('fields'), `U{` ('unnamed fields'), or `I{` ('instance'). More verbose ones I would not consider.

**Comment by catb0t** [comment]
The main complaint about `T{ foo f 1 2 }` is that the `f` switch is ugly, right? Why not change only that, replace the meaningless `f` with a meaningful word? 

One or the other of these, or maybe both for consistency? That's my 2 cents suggestion, and these would round-trip.

You don't even need to define new symbols, just use the presence of the existing `boa` or `assoc` words. 

```factor
T{ foo boa 1 2 3 } 
T{ foo assoc { a 1 } { b 2 } { c 3 } }
```

**Comment by mrjbq7** [comment]
So far my favorite is

T[ foo 1 2 3 ]

And

T{ foo { a 1 } { b 2 } { c 3 } }

Alternatively

T{ foo a: 1 b: 2 c: 3 }

But I can’t remember right now how this fit into the new proposed parser.

**Comment by alex-ilin** [comment]
Replace `f` with `|`, and you get something vaguely mathematical: `T{ foo | 1 2 3 }`
